agent/immortalstreams/backedpipe/backed_writer.go

```diff
@@ -0,0 +1,244 @@
+package backedpipe
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"golang.org/x/xerrors"
+)
+
+// BackedWriter wraps an unreliable io.Writer and makes it resilient to disconnections.
+// It maintains a ring buffer of recent writes for replay during reconnection and
+// always writes to the buffer even when disconnected.
+type BackedWriter struct {
+	mu          sync.Mutex
+	cond        *sync.Cond
+	writer      io.Writer
+	buffer      *RingBuffer
+	sequenceNum uint64 // total bytes written
+	closed      bool
+
+	// Error callback to notify parent when connection fails
+	onError func(error)
+}
+
+// NewBackedWriter creates a new BackedWriter with a 64MB ring buffer.
+// The writer is initially disconnected and will buffer writes until connected.
+func NewBackedWriter() *BackedWriter {
+	return NewBackedWriterWithCapacity(64 * 1024 * 1024)
+}
+
+// NewBackedWriterWithCapacity creates a new BackedWriter with the specified buffer capacity.
+// The writer is initially disconnected and will buffer writes until connected.
+func NewBackedWriterWithCapacity(capacity int) *BackedWriter {
+	bw := &BackedWriter{
+		buffer: NewRingBufferWithCapacity(capacity),
+	}
+	bw.cond = sync.NewCond(&bw.mu)
+	return bw
+}
+
+// Write implements io.Writer. It always writes to the ring buffer, even when disconnected.
+// When connected, it also writes to the underlying writer. If the underlying write fails,
+// the writer is marked as disconnected but the buffer write still succeeds.
+func (bw *BackedWriter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	bw.mu.Lock()
+	defer bw.mu.Unlock()
+
+	if bw.closed {
+		return 0, io.ErrClosedPipe
+	}
+
+	// Always write to buffer first
+	written, _ := bw.buffer.Write(p)
+	//nolint:gosec // Safe conversion: written is always non-negative from buffer.Write
+	bw.sequenceNum += uint64(written)
+
+	// If connected, also write to underlying writer
+	if bw.writer != nil {
+		// Unlock during actual write to avoid blocking other operations
+		bw.mu.Unlock()
+		n, err := bw.writer.Write(p)
+		bw.mu.Lock()
+
+		if n != len(p) {
+			err = xerrors.Errorf("partial write: wrote %d of %d bytes", n, len(p))
+		}
+
+		if err != nil {
+			// Connection failed, mark as disconnected
+			bw.writer = nil
+
+			// Notify parent of error if callback is set
+			if bw.onError != nil {
+				bw.onError(err)
+			}
+		}
+	}
```

Comment:
Writes need to block while we are disconnected. Otherwise, the reliable connection doesn't get any backpressure and can just sling tons of data at us that will blow out the ring buffer.